Implemented custom dialling with local address and timeout

insomniacslk opened the pull request:
Implemented custom dialling with local address and timeout as Client attributes.

This is a slightly different implementation of https://github.com/miekg/dns/pull/471 , and enables the user to set a custom local address. E.g.:

```
msg := &dns.Msg{/* something here */}
c := new(dns.Client)
c.LocalAddr = &net.UDPAddr{IP: net.ParseIP("127.0.0.1"), Port: 12345, Zone: ""}
resp, ttl, err := client.Exchange(msg, "server.example.org:53")
```

See also the related tests for usage.

insomniacslk commented:
See https://github.com/miekg/dns/issues/501 and https://github.com/miekg/dns/pull/471 for context

matje commented on client.go:
```diff
@@ -178,6 +179,28 @@ func (c *Client) ExchangeContext(ctx context.Context, m *Msg, a string) (
+	}
+	return conn, nil
+}
+
```
It probably makes sense to move these methods below the functions they are referring to, so under `DialTimeout` and `DialTimeoutWithTLS`.

insomniacslk replied:
I generally agree with you, but looking at this file's layout, I see, in order:
1) type definitions
2) various `Exchange*` functions
3) all the struct methods of `Client`
4) the various `Dial*` functions (non-struct methods)

If I move those two new `Client` methods below the referenced functions (i.e. after `4`) we would fragment the `Client` methods (`3`) along the file, making it less natural to follow the file content.

What about instead moving `Dial` and `DialTimeout` above, i.e. before any `Client` method is defined? CC @miekg as this would touch other parts of the file

matje replied:
I see your argument now for putting the methods here. Either way is fine with me.

insomniacslk replied:
I've move those two functions earlier in the file, let's see what @miekg thinks of this :-)

insomniacslk commented:
@miekg , friendly ping - can you please review this?

miekg commented:
[ Quoting <notifications@github.com> in "Re: [miekg/dns] Implemented custom ..." ]
>@miekg , friendly ping - can you please review this?

maaaaaybe tonight - busy with work on coredns atm.

miekg commented on client.go:
```diff
@@ -178,6 +242,28 @@ func (c *Client) ExchangeContext(ctx context.Context, m *Msg, a string) (
+	dialer := net.Dialer{Timeout: c.dialTimeout(), LocalAddr: c.LocalAddr}
+
+	conn = new(Conn)
+	if conn.Conn, err = tls.DialWithDialer(&dialer, network, address, c.TLSConfig); err != nil {
```
Is the `WithDialer` a thing we've fully explored?

With every new transport (granted they are not added daily), we get a new  `DialWithTLS` and `DialWithTimeoutXXX`... would it make sense to use a net.Dailer and plumb that through as "the new way"?

Code-wise this all looks fine.

insomniacslk replied:
If I understand this correctly, this is similar to what I suggested at https://github.com/miekg/dns/issues/501#issuecomment-320787105 (point 2). I am in favour of a new API that encompasses a `net.Dialer`, while keeping the old API alive but deprecated (until it gets removed).

For this specific change, I'd keep it scoped, so the refactoring could be done just after.

insomniacslk replied:
super-friendly ping :)

Once (if) this is merged I will attempt some rework of the API and submit for review

miekg commented:
circling back here. Today I finally have time to properly review and give feedback. (well later today :) )

miekg commented:
Merging this will make it part of the API and then we need to deprecate it somehow again...

As mentioned. I think we should drop this and go for a `WithDialer` option immediately.

miekg commented:
If the WithDialer code looks as clean as this one I have no problem merging that.

insomniacslk commented:
Sorry for the delay. I had some time today to think about an API redesign. What I did so far:
- created a generic method `Client.dialWithDialer` that encompasses timeout, local address, network and whether to use TLS or not (return values are unchanged);
- modified `Client.exchange` (the unexported one) to take as input a context, a Msg and a resolver address (return values are unchanged);
- all the rest of the API is unchanged, but some functions/methods are deprecated and moved to a `client_compat.go` file to make it easy to distinguish the old and new API.

I will also add a crapload of tests to cover as many cases as possible.

I have a working prototype that does what said above (except the test coverage). What do you think of this plan? To be clear, the existing API is unchanged and in some parts deprecated. Such existing API is reimplemented on top of the new API.

One thing of note: Exchange and ExchangeContext behave differently despite what documented (the latter honours `SingleInflight`, the former and all the rest of the code don't). Shall I backport this feature to Exchange (and similarly named functions/methods) or not?

insomniacslk commented:
that actually makes sense, as `Exchange` is not attached to any client, but `SingleInflight` is client-specific. I'll keep the current behaviour

miekg commented:
Yes this sounds good!

insomniacslk commented:
Note to self: also update `doc.go` :)

insomniacslk commented:
Sorry, holidays and a small mountain accident kept me away from this.

@miekg I have one doubt on the design where I'd like your input.

Background: there are several ways to set a timeout, and using a `net.Dialer` adds one more. At the moment there are:
* dial timeout
* read timeout
* write timeout
* client timeout (overrides all the previous ones)
* context timeout (overrides client timeout)

With the dialer, we are adding a new timeout (`net.Dialer.Timeout`) which, just like the context timeout, should override any other timeout of the ones above (except context timeout, but just because context and dialer are never used together).

I am also for deprecating `ExchangeContext` in favour of `Client.ExchangeWithDialer` and `Client.DialWithDialer` (and the existing methods are just wrappers around it), since it is only used for timeouts, and I agree with Dave Cheney that context should not be used for cancellation either, https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation .

My proposed solution on this specific topic is to:
* deprecate (but not remove) `ExchangeContext` and other `*Context` methods, and wrap them around the `*WithDialer`. Context is only used for timeout and I don't like the idea of carrying around both a Context and a Dialer
* context timeout becomes dialer timeout (i.e. the wrapper takes the context timeout but then uses a dialer under the hood)
* dialer timeout overrides all the other timeouts, if present, just like the context timeout does now (this helps maintain the current behaviour in the existing API)

Does this plan make sense or do you have a preference to keep `*Context` methods around?

insomniacslk commented:
And another question for @miekg : what do you think of using GoMock (https://github.com/golang/mock) for better testing? This would allow us to mock tls.DialWithDialer, net.Dialer.Dial and other library functions so I can do proper unit testing in addition to integration tests

miekg commented:
(also on holiday)

I don't like those mocking frameworks. Also adds a non stdlib dependency.
Of course not hanging those, actually testing those will require more code.

Re: your other question: assume net.Dialer timeouts take precedence over
all others.

On Aug 29, 2017 20:49, "insomniac" <notifications@github.com> wrote:

And another question for @miekg <https://github.com/miekg> : what do you
think of using GoMock (https://github.com/golang/mock) for better testing?
This would allow us to mock tls.DialWithDialer, net.Dialer.Dial and other
library functions so I can do proper unit testing in addition to
integration tests

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/miekg/dns/pull/502#issuecomment-325759403>, or mute the
thread
<https://github.com/notifications/unsubscribe-auth/AAVkW1GNKAR4SHwB6QSDA7Pe1zr6SfUKks5sdF07gaJpZM4OwD_j>
.

insomniacslk commented:
I have created https://github.com/miekg/dns/pull/511 since it's a larger change than the one here (that I will close)

miekg commented:
closing favor of #511
